linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,346 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <semaphore.h>
+#include <time.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/gballoc_hl.h" // IWYU pragma: keep
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS          100
+#define MAX_THREAD_COUNT    1
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+    struct THREADPOOL_TASK_TAG* next;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    volatile_atomic int32_t pool_idle;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Queue
+    THREADPOOL_TASK* task_list;
+    THREADPOOL_TASK* task_list_last;
+
+    SRW_LOCK_HANDLE srw_lock;
+
+    THREAD_HANDLE* thread_handle_list;
+    uint32_t list_index;
+} THREADPOOL;
+
+int threadpool_work_func(void* param)
+{
+    if (param == NULL)
+    {
+        LogCritical("Invalid args: param: %p", param);
+    }
+    else
+    {
+        struct timespec ts;
+        THREADPOOL* threadpool = param;
+        do
+        {
+            (void)interlocked_increment(&threadpool->pool_idle);
```

Comment:
>pool_idle [](http://example.com/codeflow?start=53&length=9)

Doesn't the semaphore do this counting for us? As in, shouldn't the semaphore be keeping a count of the number of idle threads (which is incremented by these threads waiting and decremented by signaling the semaphore)?

Maybe I didn't think through some other case as we are going from 1 to MAX threads, but I would think we shouldn't need both an idle count and a semaphore.

Reply:
I don't think we need the idle counter.